### PR TITLE
Tips notifications for rarely used functionality

### PR DESCRIPTION
**Pull request by etatanova**
**Related issue:** #1278 

_Changes:_
- Implemented tips notifications logic
- Added tips storage
- Moved the logic of downloading a config to a separate module

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -52,6 +52,7 @@ Using this extension, you can **debug your code and quickly run `react-native` c
   - [Change project root](#change-project-root)
   - [Configure an Android LogCat Monitor](#configure-an-android-logcat-monitor)
   - [Configure dependencies versions for debugging Expo projects](#configure-dependencies-versions-for-debugging-expo-projects)
+  - [Configure the same custom keybindings for different commands](#configure-the-same-custom-keybindings-for-different-commands)
```
Comment (RedMickey) [inline README.md:55]:
```suggestion
  - [Configure custom key bindings for extension commands](#configure-custom-key-bindings-for-extension-commands)
```

File: README.md [hunk 2]
```diff
@@ -731,6 +743,10 @@ Before using the Network inspector, please make sure that your system meets the
 
 For now the Network inspector doesn't support Expo applications.
 
+### Network inspector logs theme
+
+The extension provides “Dark” and “Light” color themes for Network Inspector logs. You can configure the theme in the extension configuration in the Settings tab or in setting.json file.
```
Comment (RedMickey) [inline README.md:748]:
```suggestion
The extension provides “Dark” and “Light” color themes for Network Inspector logs. You can configure the theme in the extension configuration in the Settings tab or in `settings.json` file.
```

File: src/extension/tipsNotificationsService/tipsNotificationService.ts [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+export class TipNotificationService implements vscode.Disposable {
+    private static instance: TipNotificationService;
+
+    private readonly endpointURL: string;
+    private cancellationTokenSource: vscode.CancellationTokenSource;
+
+    private readonly downloadConfigRequest: Promise<TipNotificationConfig>;
+    private logger: OutputChannelLogger;
+
+    private TIPS_NOTIFICATIONS_LOG_CHANNEL_NAME = "Tips Notifications";
```
Comment (RedMickey) [inline src/extension/tipsNotificationsService/tipsNotificationService.ts:64]:
```suggestion
    private readonly TIPS_NOTIFICATIONS_LOG_CHANNEL_NAME = "Tips Notifications";
```

File: src/extension/tipsNotificationsService/tipsNotificationService.ts [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+        return tipsConfig;
+    }
+
+    public async showTipNotification(
+        isGeneralTip: boolean = true,
+        specificTipKey?: string,
+    ): Promise<void> {
+        if (!isGeneralTip && !specificTipKey) {
+            this.logger.debug(
+                "For specific tips, it is necessary to pass the tip key as a parameter",
```
Comment (RedMickey) [inline src/extension/tipsNotificationsService/tipsNotificationService.ts:131]:
```suggestion
                "The specific tip key parameter isn't passed for a specific tip",
```

File: src/extension/remoteConfigHelper.ts [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+                    try {
+                        resolve(JSON.parse(data));
+                    } catch (err) {
+                        reject(err);
+                    }
+                });
+                response.on("error", reject);
+            })
+            .on("error", reject);
+    });
```
Comment (JiglioNero) [inline src/extension/remoteConfigHelper.ts:27]: There is already implemented the same [function](https://github.com/microsoft/vscode-react-native/blob/master/src/common/node/request.ts#L7)

File: src/extension/remoteConfigHelper.ts [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    try {
+        return await downloadConfig<T>(endpointURL);
+    } catch (err) {
+        if (retryCount < 1 || cancellationTokenSource.token.isCancellationRequested) {
+            throw err;
+        }
+
+        await PromiseUtil.delay(2000);
+        return await retryDownloadConfig(endpointURL, cancellationTokenSource, --retryCount);
+    }
```
Comment (JiglioNero) [inline src/extension/remoteConfigHelper.ts:44]: There is already implemented the same [function](https://github.com/microsoft/vscode-react-native/blob/master/src/common/node/promise.ts#L26-L34)

File: src/extension/tipsNotificationsService/tipsNotificationService.ts [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
+        );
+        this.logger = OutputChannelLogger.getChannel(this.TIPS_NOTIFICATIONS_LOG_CHANNEL_NAME);
+    }
+
+    private async getOrCreateDefaultTipsConfig(): Promise<TipsConfig> {
+        if (ExtensionConfigManager.config.has("tipsConfig")) {
+            return ExtensionConfigManager.config.get("tipsConfig");
+        }
+
+        const tipNotificationService: TipNotificationService = TipNotificationService.getInstance();
```
Comment (JiglioNero) [inline src/extension/tipsNotificationsService/tipsNotificationService.ts:97]: I think there we can just use `this` instead of `getInstance()` function

**Review by JiglioNero — APPROVED** [review]

**Review by RedMickey — APPROVED** [review]

**Pull request merged**